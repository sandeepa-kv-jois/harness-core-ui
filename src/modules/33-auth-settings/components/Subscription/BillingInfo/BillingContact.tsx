/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { Card, Layout, Text, TextInput, DropDown, FontVariation, Color } from '@harness/uicore'
import { useStrings } from 'framework/strings'
import type { BillingContactProps } from '@common/constants/SubscriptionTypes'
import css from './BillingInfo.module.scss'

interface BillingContactProp {
  billingInfo: BillingContactProps
  setBillingInfo: (value: BillingContactProps) => void
}

const BillingContact: React.FC<BillingContactProp> = ({ billingInfo, setBillingInfo }) => {
  const { getString } = useStrings()
  const { companyName, email, country, billingAddress, city, state, zipCode } = billingInfo
  return (
    <Card>
      <Layout.Vertical spacing={'large'}>
        <Text font={{ variation: FontVariation.H5 }}>{getString('authSettings.billingInfo.billingContact')}</Text>
        <Layout.Vertical spacing={'large'}>
          <Layout.Horizontal>
            <Layout.Vertical width={'30%'} spacing="small" padding={{ right: 'large' }}>
              <Text>{getString('common.headerCompanyName')}</Text>
              <TextInput
                data-testid="companyName"
                value={companyName}
                onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
                  setBillingInfo({
                    ...billingInfo,
                    companyName: e.target.value
                  })
                }}
              />
            </Layout.Vertical>
            <Layout.Vertical width={'70%'}>
              <Text padding={{ bottom: 'small' }}>{getString('email')}</Text>
              <TextInput
                data-testid="email"
                value={email}
                onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
                  setBillingInfo({
                    ...billingInfo,
                    email: e.target.value
                  })
                }}
              />
              <Text
                icon="info"
                iconProps={{ color: Color.PRIMARY_7 }}
                padding="small"
                margin={{ bottom: 'small' }}
                className={css.warning}
              >
                {getString('authSettings.billingInfo.emailWarning')}
              </Text>
            </Layout.Vertical>
          </Layout.Horizontal>
          <Layout.Horizontal>
            <Layout.Vertical width={'30%'} spacing="small" padding={{ right: 'large' }}>
              <Text>{getString('common.countryOrRegion')}</Text>
              <DropDown
                buttonTestId="countryOrRegion"
                items={[{ label: 'US', value: 'us' }]}
                value={country}
                onChange={option => {
                  setBillingInfo({
                    ...billingInfo,
                    country: String(option.value)
                  })
                }}
              />
            </Layout.Vertical>
            <Layout.Vertical width={'70%'}>
              <Text padding={{ bottom: 'small' }}>{getString('common.billingAddress')}</Text>
              <TextInput
                data-testid="billingAddress"
                value={billingAddress}
                onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
                  setBillingInfo({
                    ...billingInfo,
                    billingAddress: e.target.value
                  })
                }}
              />
            </Layout.Vertical>
          </Layout.Horizontal>
          <Layout.Horizontal>
            <Layout.Vertical width={'30%'} spacing="small" padding={{ right: 'large' }}>
              <Text>{getString('common.city')}</Text>
              <TextInput
                data-testid="city"
                value={city}
                onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
                  setBillingInfo({
                    ...billingInfo,
                    city: e.target.value
                  })
                }}
              />
            </Layout.Vertical>
            <Layout.Horizontal flex={{ justifyContent: 'space-between', alignItems: 'baseline' }} width={'70%'}>
              <Layout.Vertical spacing="small">
                <Text>{getString('common.state')}</Text>
                <DropDown
                  buttonTestId="state"
                  items={[{ label: 'Texas', value: 'Texas' }]}
                  value={state}
                  onChange={option => {
                    setBillingInfo({
                      ...billingInfo,
                      state: String(option.value)
                    })
                  }}
                />
              </Layout.Vertical>
              <Layout.Vertical spacing="small">
                <Text>{getString('common.zipCode')}</Text>
                <TextInput
                  data-testid="zipCode"
                  value={zipCode}
                  onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
                    setBillingInfo({
                      ...billingInfo,
                      zipCode: e.target.value
                    })
                  }}
                />
              </Layout.Vertical>
            </Layout.Horizontal>
          </Layout.Horizontal>
        </Layout.Vertical>
      </Layout.Vertical>
    </Card>
  )
}

export default BillingContact
