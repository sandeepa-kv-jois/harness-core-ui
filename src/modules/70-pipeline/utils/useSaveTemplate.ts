/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { cloneDeep, defaultTo, isEmpty, omit } from 'lodash-es'
import { parse } from 'yaml'
import { useHistory, useParams } from 'react-router-dom'
import { VisualYamlSelectedView as SelectedView } from '@wings-software/uicore'
import {
  createTemplatePromise,
  EntityGitDetails,
  NGTemplateInfoConfig,
  TemplateSummaryResponse,
  updateExistingTemplateLabelPromise
} from 'services/template-ng'
import { AppStoreContext } from 'framework/AppStore/AppStoreContext'
import { useStrings } from 'framework/strings'
import useRBACError, { RBACError } from '@rbac/utils/useRBACError/useRBACError'
import { useToaster } from '@common/exports'
import { UseSaveSuccessResponse, useSaveToGitDialog } from '@common/modals/SaveToGitDialog/useSaveToGitDialog'
import type { SaveToGitFormInterface } from '@common/components/SaveToGitForm/SaveToGitForm'
import { DefaultNewTemplateId } from 'framework/Templates/templates'
import { yamlStringify } from '@common/utils/YamlHelperMethods'
import { sanitize } from '@common/utils/JSONUtils'
import routes from '@common/RouteDefinitions'
import type { GitQueryParams, ModulePathParams, TemplateStudioPathProps } from '@common/interfaces/RouteInterfaces'
import { useQueryParams } from '@common/hooks'
import type { PromiseExtraArgs } from 'framework/Templates/TemplateConfigModal/TemplateConfigModal'
import type { YamlBuilderHandlerBinding } from '@common/interfaces/YAMLBuilderProps'
import { getScopeFromDTO } from '@common/components/EntityReference/EntityReference'
import { Scope } from '@common/interfaces/SecretsInterface'
import { StoreMetadata, StoreType } from '@common/constants/GitSyncTypes'

export interface FetchTemplateUnboundProps {
  forceFetch?: boolean
  forceUpdate?: boolean
  signal?: AbortSignal
  repoIdentifier?: string
  branch?: string
}

declare global {
  interface WindowEventMap {
    TEMPLATE_SAVED: CustomEvent<TemplateSummaryResponse>
  }
}

interface SaveTemplateObj {
  template: NGTemplateInfoConfig
}

interface UseSaveTemplateReturnType {
  saveAndPublish: (
    updatedTemplate: NGTemplateInfoConfig,
    extraInfo: PromiseExtraArgs
  ) => Promise<UseSaveSuccessResponse>
}

export interface TemplateContextMetadata {
  yamlHandler?: YamlBuilderHandlerBinding
  gitDetails?: EntityGitDetails
  storeMetadata?: StoreMetadata
  setLoading?: (loading: boolean) => void
  fetchTemplate?: (args: FetchTemplateUnboundProps) => Promise<void>
  deleteTemplateCache?: (gitDetails?: EntityGitDetails) => Promise<void>
  view?: string
  isPipelineStudio?: boolean
  stableVersion?: string
  fireSuccessEvent?: boolean
}

export function useSaveTemplate(TemplateContextMetadata: TemplateContextMetadata): UseSaveTemplateReturnType {
  const {
    yamlHandler,
    gitDetails,
    storeMetadata,
    setLoading,
    fetchTemplate,
    deleteTemplateCache,
    view,
    isPipelineStudio,
    stableVersion,
    fireSuccessEvent
  } = TemplateContextMetadata
  const { isGitSyncEnabled } = React.useContext(AppStoreContext)
  const { templateIdentifier, templateType, projectIdentifier, orgIdentifier, accountId, module } = useParams<
    TemplateStudioPathProps & ModulePathParams
  >()
  const { branch } = useQueryParams<GitQueryParams>()
  const { getString } = useStrings()
  const { showSuccess, showError, clear } = useToaster()
  const { getRBACErrorMessage } = useRBACError()
  const history = useHistory()
  const isYaml = view === SelectedView.YAML

  const navigateToLocation = (newTemplate: NGTemplateInfoConfig, updatedGitDetails?: SaveToGitFormInterface): void => {
    history.replace(
      routes.toTemplateStudio({
        projectIdentifier: newTemplate.projectIdentifier,
        orgIdentifier: newTemplate.orgIdentifier,
        accountId,
        ...(!isEmpty(newTemplate.projectIdentifier) && { module }),
        templateType: templateType,
        templateIdentifier: newTemplate.identifier,
        versionLabel: newTemplate.versionLabel,
        repoIdentifier: updatedGitDetails?.repoIdentifier,
        branch: updatedGitDetails?.branch
      })
    )
  }

  const stringifyTemplate = React.useCallback(
    // Important to sanitize the final template to avoid sending null values as it fails schema validation
    (temp: NGTemplateInfoConfig) =>
      yamlStringify(
        JSON.parse(
          JSON.stringify({
            template: sanitize(temp, {
              removeEmptyString: false,
              removeEmptyObject: false,
              removeEmptyArray: false
            })
          })
        ),
        {
          version: '1.1'
        }
      ),
    []
  )

  const updateExistingLabel = async (
    latestTemplate: NGTemplateInfoConfig,
    comments?: string,
    updatedGitDetails?: SaveToGitFormInterface,
    lastObject?: { lastObjectId?: string },
    currStoreMetadata?: StoreMetadata
  ): Promise<UseSaveSuccessResponse> => {
    if (storeMetadata) {
      currStoreMetadata = storeMetadata
    }

    try {
      const response = await updateExistingTemplateLabelPromise({
        templateIdentifier: latestTemplate.identifier,
        versionLabel: latestTemplate.versionLabel,
        body: stringifyTemplate(latestTemplate),
        queryParams: {
          accountIdentifier: accountId,
          projectIdentifier,
          orgIdentifier,
          comments,
          ...(updatedGitDetails ?? {}),
          ...(lastObject?.lastObjectId ? lastObject : {}),
          ...(updatedGitDetails && updatedGitDetails.isNewBranch ? { baseBranch: branch } : {}),
          ...(currStoreMetadata?.storeType === StoreType.REMOTE ? currStoreMetadata : {})
        },
        requestOptions: { headers: { 'Content-Type': 'application/yaml' } }
      })
      setLoading?.(false)
      if (response && response.status === 'SUCCESS') {
        if (!isGitSyncEnabled && currStoreMetadata?.storeType !== StoreType.REMOTE) {
          clear()
          showSuccess(getString('common.template.updateTemplate.templateUpdated'))
        }
        await fetchTemplate?.({ forceFetch: true, forceUpdate: true })
        if (updatedGitDetails?.isNewBranch) {
          navigateToLocation(latestTemplate, updatedGitDetails)
        }
        return { status: response.status }
      } else {
        throw response
      }
    } catch (error) {
      clear()
      if (!isGitSyncEnabled && currStoreMetadata?.storeType !== StoreType.REMOTE) {
        showError(getRBACErrorMessage(error as RBACError), undefined, 'template.update.template.error')
        return { status: 'FAILURE' }
      } else {
        throw error
      }
    }
  }

  const saveAndPublishTemplate = async (
    latestTemplate: NGTemplateInfoConfig,
    comments?: string,
    isEdit = false,
    updatedGitDetails?: SaveToGitFormInterface,
    lastObject?: { lastObjectId?: string },
    currStoreMetadata?: StoreMetadata
  ): Promise<UseSaveSuccessResponse> => {
    const isGitExperienceEnabled = isGitSyncEnabled && getScopeFromDTO(latestTemplate) === Scope.PROJECT
    if (storeMetadata) {
      currStoreMetadata = storeMetadata
    }

    if (!isGitExperienceEnabled && currStoreMetadata?.storeType !== StoreType.REMOTE) {
      setLoading?.(true)
    }

    if (isEdit) {
      return updateExistingLabel(latestTemplate, comments, updatedGitDetails, lastObject, currStoreMetadata)
    } else {
      try {
        const response = await createTemplatePromise({
          body: stringifyTemplate(omit(cloneDeep(latestTemplate), 'repo', 'branch')),
          queryParams: {
            accountIdentifier: accountId,
            projectIdentifier: latestTemplate.projectIdentifier,
            orgIdentifier: latestTemplate.orgIdentifier,
            comments,
            ...(updatedGitDetails ?? {}),
            ...(updatedGitDetails && updatedGitDetails.isNewBranch ? { baseBranch: branch } : {}),
            ...(currStoreMetadata?.storeType === StoreType.REMOTE ? currStoreMetadata : {})
          },
          requestOptions: { headers: { 'Content-Type': 'application/yaml' } }
        })
        if (!isGitExperienceEnabled && currStoreMetadata?.storeType !== StoreType.REMOTE) {
          setLoading?.(false)
        }
        if (response && response.status === 'SUCCESS') {
          if (fireSuccessEvent && response.data?.templateResponseDTO) {
            window.dispatchEvent(new CustomEvent('TEMPLATE_SAVED', { detail: response.data.templateResponseDTO }))
          }
          if (!isGitExperienceEnabled && currStoreMetadata?.storeType !== StoreType.REMOTE) {
            clear()
            showSuccess(getString('common.template.saveTemplate.publishTemplate'))
          }
          await deleteTemplateCache?.()
          if (!isPipelineStudio) {
            navigateToLocation(latestTemplate, updatedGitDetails)
          }
          return { status: response.status }
        } else {
          throw response
        }
      } catch (error) {
        clear()
        if (!isGitExperienceEnabled && currStoreMetadata?.storeType !== StoreType.REMOTE) {
          showError(getRBACErrorMessage(error as RBACError), undefined, 'template.save.template.error')
          return { status: 'FAILURE' }
        } else {
          throw error
        }
      }
    }
  }

  const saveAndPublishWithGitInfo = async (
    updatedGitDetails: SaveToGitFormInterface,
    payload?: SaveTemplateObj,
    objectId?: string,
    isEdit = false,
    currStoreMetadata?: StoreMetadata
  ): Promise<UseSaveSuccessResponse> => {
    let latestTemplate = payload?.template as NGTemplateInfoConfig

    if (isYaml && yamlHandler) {
      try {
        latestTemplate = payload?.template || (parse(yamlHandler.getLatestYaml()).pipeline as NGTemplateInfoConfig)
      } /* istanbul ignore next */ catch (err) {
        showError(getRBACErrorMessage(err as RBACError), undefined, 'template.save.gitinfo.error')
      }
    }

    const response = await saveAndPublishTemplate(
      latestTemplate,
      '',
      isEdit,
      omit(updatedGitDetails, 'name', 'identifier'),
      templateIdentifier !== DefaultNewTemplateId ? { lastObjectId: objectId } : {},
      currStoreMetadata
    )
    return {
      status: response?.status
    }
  }

  const { openSaveToGitDialog } = useSaveToGitDialog<SaveTemplateObj>({
    onSuccess: (
      gitData: SaveToGitFormInterface,
      payload?: SaveTemplateObj,
      objectId?: string,
      isEdit = false,
      currStoreMetadata?: StoreMetadata
    ): Promise<UseSaveSuccessResponse> =>
      saveAndPublishWithGitInfo(gitData, payload, defaultTo(objectId, ''), isEdit, currStoreMetadata)
  })

  const getUpdatedGitDetails = (
    currGitDetails: EntityGitDetails,
    latestTemplate: NGTemplateInfoConfig,
    isEdit: boolean | undefined = false
  ): EntityGitDetails => ({
    ...currGitDetails,
    ...(!isEdit && {
      filePath: `${defaultTo(latestTemplate.identifier, '')}_${defaultTo(latestTemplate.versionLabel, '').replace(
        /[^a-zA-Z0-9-_]/g,
        ''
      )}.yaml`
    })
  })

  const saveAndPublish = React.useCallback(
    async (updatedTemplate: NGTemplateInfoConfig, extraInfo: PromiseExtraArgs): Promise<UseSaveSuccessResponse> => {
      let currStoreMetadata = storeMetadata
      if (!currStoreMetadata) {
        currStoreMetadata = extraInfo?.storeMetadata
      }

      const { isEdit, comment, updatedGitDetails } = extraInfo

      // if Git sync enabled then display modal
      if (isGitSyncEnabled || currStoreMetadata?.storeType === StoreType.REMOTE) {
        if (
          (currStoreMetadata?.storeType !== StoreType.REMOTE && isEmpty(gitDetails?.repoIdentifier)) ||
          isEmpty(gitDetails?.branch)
        ) {
          clear()
          showError(getString('pipeline.gitExperience.selectRepoBranch'))
          return Promise.reject(getString('pipeline.gitExperience.selectRepoBranch'))
        } else if (updatedGitDetails) {
          openSaveToGitDialog({
            isEditing: defaultTo(isEdit, false),
            resource: {
              type: 'Template',
              name: updatedTemplate.name,
              identifier: updatedTemplate.identifier,
              gitDetails: getUpdatedGitDetails(updatedGitDetails, updatedTemplate, isEdit),
              storeMetadata: currStoreMetadata
            },
            payload: { template: omit(updatedTemplate, 'repo', 'branch') }
          })
          return Promise.resolve({ status: 'SUCCESS' })
        }
      } else {
        return saveAndPublishTemplate(updatedTemplate, comment, isEdit)
      }
    },
    [
      storeMetadata,
      isGitSyncEnabled,
      gitDetails?.repoIdentifier,
      gitDetails?.branch,
      clear,
      showError,
      getString,
      openSaveToGitDialog,
      saveAndPublishTemplate
    ]
  )

  return {
    saveAndPublish
  }
}
