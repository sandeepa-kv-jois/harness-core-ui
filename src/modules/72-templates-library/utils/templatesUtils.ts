/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { isEmpty } from 'lodash-es'
import type { UseStringsReturn } from 'framework/strings'
import type { TemplateSummaryResponse } from 'services/template-ng'

export enum TemplateType {
  Step = 'Step',
  Stage = 'Stage',
  Pipeline = 'Pipeline',
  Service = 'Service',
  Infrastructure = 'Infrastructure',
  StepGroup = 'StepGroup',
  Execution = 'Execution',
  MonitoredService = 'MonitoredService',
  SecretManager = 'SecretManager'
}

export const getAllowedTemplateTypes = (
  getString: UseStringsReturn['getString'],
  module?: string,
  scriptTemplateEnabled?: boolean
): { label: string; value: string; disabled?: boolean }[] => {
  const AllowedTemplateTypes = [
    {
      label: getString('step'),
      value: TemplateType.Step,
      disabled: false
    },
    {
      label: getString('common.stage'),
      value: TemplateType.Stage,
      disabled: false
    },
    {
      label: getString('common.pipeline'),
      value: TemplateType.Pipeline,
      disabled: false
    },
    {
      label: getString('service'),
      value: TemplateType.Service,
      disabled: true
    },
    {
      label: getString('infrastructureText'),
      value: TemplateType.Infrastructure,
      disabled: true
    },
    {
      label: getString('stepGroup'),
      value: TemplateType.StepGroup,
      disabled: true
    },
    {
      label: getString('executionText'),
      value: TemplateType.Execution,
      disabled: true
    },
    ...(scriptTemplateEnabled
      ? [
          {
            label: getString('script'),
            value: TemplateType.SecretManager,
            disabled: !scriptTemplateEnabled
          }
        ]
      : [])
  ]
  if (module === 'cv') {
    return [
      {
        label: getString('connectors.cdng.monitoredService.label'),
        value: TemplateType.MonitoredService,
        disabled: false
      }
    ]
  }
  return AllowedTemplateTypes
}

export const getVersionLabelText = (template: TemplateSummaryResponse, getString: UseStringsReturn['getString']) => {
  return isEmpty(template.versionLabel)
    ? getString('templatesLibrary.alwaysUseStableVersion')
    : template.stableTemplate
    ? getString('templatesLibrary.stableVersion', { entity: template.versionLabel })
    : template.versionLabel
}

export const getTemplateRuntimeInputsCount = (templateInfo: { [key: string]: any }): number =>
  (JSON.stringify(templateInfo).match(/<\+input>/g) || []).length
