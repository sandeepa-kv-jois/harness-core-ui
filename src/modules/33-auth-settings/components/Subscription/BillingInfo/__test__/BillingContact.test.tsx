/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { TestWrapper } from '@common/utils/testUtils'
import BillingContact from '../BillingContact'

describe('BillingContact', () => {
  const billingInfo = {
    name: 'Jane Doe',
    email: 'harness@harness.io',
    country: 'us',
    zipCode: '12345',
    billingAddress: 'billing address 1',
    city: 'dallas',
    state: 'texas',
    companyName: 'Harness'
  }

  const setBillingInfoMock = jest.fn()

  test('render', async () => {
    const { container } = render(
      <TestWrapper>
        <BillingContact setBillingInfo={jest.fn()} billingInfo={billingInfo} />
      </TestWrapper>
    )
    await waitFor(() => {
      expect(container).toMatchSnapshot()
    })
  })

  test('set companyName', async () => {
    const { getByTestId } = render(
      <TestWrapper>
        <BillingContact setBillingInfo={setBillingInfoMock} billingInfo={billingInfo} />
      </TestWrapper>
    )
    userEvent.clear(getByTestId('companyName'))
    userEvent.type(getByTestId('companyName'), 'another company')
    await waitFor(() => {
      expect(setBillingInfoMock).toHaveBeenCalledWith({
        ...billingInfo,
        companyName: 'another company'
      })
    })
  })

  test('set email', async () => {
    const { getByTestId } = render(
      <TestWrapper>
        <BillingContact setBillingInfo={setBillingInfoMock} billingInfo={billingInfo} />
      </TestWrapper>
    )
    userEvent.clear(getByTestId('email'))
    userEvent.type(getByTestId('email'), 'new@email.com')
    await waitFor(() => {
      expect(setBillingInfoMock).toHaveBeenCalledWith({
        ...billingInfo,
        email: 'new@email.com'
      })
    })
  })

  test('set zipCode', async () => {
    const { getByTestId } = render(
      <TestWrapper>
        <BillingContact setBillingInfo={setBillingInfoMock} billingInfo={billingInfo} />
      </TestWrapper>
    )
    userEvent.clear(getByTestId('zipCode'))
    userEvent.type(getByTestId('zipCode'), '54321')
    await waitFor(() => {
      expect(setBillingInfoMock).toHaveBeenCalledWith({
        ...billingInfo,
        zipCode: '54321'
      })
    })
  })

  test('set billingAddress', async () => {
    const { getByTestId } = render(
      <TestWrapper>
        <BillingContact setBillingInfo={setBillingInfoMock} billingInfo={billingInfo} />
      </TestWrapper>
    )
    userEvent.clear(getByTestId('billingAddress'))
    userEvent.type(getByTestId('billingAddress'), 'new billing address')
    await waitFor(() => {
      expect(setBillingInfoMock).toHaveBeenCalledWith({
        ...billingInfo,
        billingAddress: 'new billing address'
      })
    })
  })

  test('set city', async () => {
    const { getByTestId } = render(
      <TestWrapper>
        <BillingContact setBillingInfo={setBillingInfoMock} billingInfo={billingInfo} />
      </TestWrapper>
    )
    userEvent.clear(getByTestId('city'))
    userEvent.type(getByTestId('city'), 'austin')
    await waitFor(() => {
      expect(setBillingInfoMock).toHaveBeenCalledWith({
        ...billingInfo,
        city: 'austin'
      })
    })
  })

  test('set state', async () => {
    const { getByTestId, container } = render(
      <TestWrapper>
        <BillingContact setBillingInfo={setBillingInfoMock} billingInfo={billingInfo} />
      </TestWrapper>
    )

    const dropdown = getByTestId('state')
    fireEvent.click(dropdown)

    const listItem = container.getElementsByClassName('DropDown--menuItem')[0]
    fireEvent.click(listItem)
    await waitFor(() => {
      expect(setBillingInfoMock).toHaveBeenCalledWith({
        ...billingInfo,
        state: 'Texas'
      })
    })
  })

  test('set countryOrRegion', async () => {
    const { getByTestId, container } = render(
      <TestWrapper>
        <BillingContact setBillingInfo={setBillingInfoMock} billingInfo={billingInfo} />
      </TestWrapper>
    )

    const dropdown = getByTestId('countryOrRegion')
    fireEvent.click(dropdown)

    const listItem = container.getElementsByClassName('DropDown--menuItem')[0]
    fireEvent.click(listItem)
    await waitFor(() => {
      expect(setBillingInfoMock).toHaveBeenCalledWith({
        ...billingInfo,
        country: 'us'
      })
    })
  })
})
