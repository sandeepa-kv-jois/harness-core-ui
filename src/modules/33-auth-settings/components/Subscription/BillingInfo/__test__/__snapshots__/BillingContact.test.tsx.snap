// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BillingContact render 1`] = `
<div>
  <div
    class="bp3-card bp3-elevation-0 Card--card"
  >
    <div
      class="Layout--vertical Layout--layout-spacing-large StyledProps--main"
    >
      <p
        class="StyledProps--font StyledProps--main StyledProps--font-variation-h5"
      >
        authSettings.billingInfo.billingContact
      </p>
      <div
        class="Layout--vertical Layout--layout-spacing-large StyledProps--main"
      >
        <div
          class="Layout--horizontal StyledProps--main"
        >
          <div
            class="Layout--vertical Layout--layout-spacing-small StyledProps--main StyledProps--padding-right-large"
            style="width: 30%;"
          >
            <p
              class="StyledProps--font StyledProps--main"
            >
              common.headerCompanyName
            </p>
            <div
              class="TextInput--main"
            >
              <div
                class="bp3-input-group"
              >
                <input
                  class="bp3-input"
                  data-testid="companyName"
                  type="text"
                  value="Harness"
                />
              </div>
            </div>
          </div>
          <div
            class="Layout--vertical StyledProps--main"
            style="width: 70%;"
          >
            <p
              class="StyledProps--font StyledProps--main StyledProps--padding-bottom-small"
            >
              email
            </p>
            <div
              class="TextInput--main"
            >
              <div
                class="bp3-input-group"
              >
                <input
                  class="bp3-input"
                  data-testid="email"
                  type="text"
                  value="harness@harness.io"
                />
              </div>
            </div>
            <p
              class="StyledProps--font StyledProps--main warning StyledProps--padding-small StyledProps--margin-bottom-small"
              style="display: flex; align-items: center;"
            >
              <span
                class="bp3-icon StyledProps--main Text--icon StyledProps--padding-right-xsmall StyledProps--color StyledProps--color-primary7"
                data-icon="info"
              >
                <svg
                  height="16"
                  viewBox="0 0 12 12"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M0 6a6 6 0 1012 0A6 6 0 000 6zm.75 0a5.25 5.25 0 1010.5 0A5.25 5.25 0 00.75 6zm4.206 3.104v.646h2.682v-.646l-.836-.144V4.203H4.956v.65l.836.144V8.96l-.836.144zm1.846-7.352v1.03h-1.01v-1.03h1.01z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
              authSettings.billingInfo.emailWarning
            </p>
          </div>
        </div>
        <div
          class="Layout--horizontal StyledProps--main"
        >
          <div
            class="Layout--vertical Layout--layout-spacing-small StyledProps--main StyledProps--padding-right-large"
            style="width: 30%;"
          >
            <p
              class="StyledProps--font StyledProps--main"
            >
              common.countryOrRegion
            </p>
            <div
              class="bp3-popover-wrapper DropDown--main"
            >
              <div
                class="bp3-popover-target"
              >
                <div
                  class=""
                >
                  <div
                    class="Layout--horizontal StyledProps--main DropDown--dropdownButton DropDown--withBorder DropDown--minWidth StyledProps--flex"
                    data-testid="countryOrRegion"
                    style="min-width: 130px;"
                  >
                    <div
                      class="Layout--horizontal StyledProps--main DropDown--labelWrapper StyledProps--flex"
                    >
                      <p
                        class="StyledProps--font Text--lineclamp StyledProps--main DropDown--label"
                        data-testid="dropdown-value"
                        style="--text-line-clamp: 1;"
                      >
                        US
                      </p>
                    </div>
                    <div
                      class="Layout--horizontal StyledProps--main DropDown--btnWrapper StyledProps--flex"
                    >
                      <span
                        class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-grey400"
                        data-icon="main-chevron-down"
                      >
                        <svg
                          height="8"
                          viewBox="0 0 14 8"
                          width="8"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M13 0c-.3 0-.5.1-.7.3L7 5.6 1.7.3C1.5.1 1.3 0 1 0 .4 0 0 .4 0 1c0 .3.1.5.3.7l6 6c.2.2.4.3.7.3s.5-.1.7-.3l6-6c.2-.2.3-.4.3-.7 0-.6-.4-1-1-1z"
                            fill="#4d5969"
                          />
                        </svg>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="Layout--vertical StyledProps--main"
            style="width: 70%;"
          >
            <p
              class="StyledProps--font StyledProps--main StyledProps--padding-bottom-small"
            >
              common.billingAddress
            </p>
            <div
              class="TextInput--main"
            >
              <div
                class="bp3-input-group"
              >
                <input
                  class="bp3-input"
                  data-testid="billingAddress"
                  type="text"
                  value="billing address 1"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="Layout--horizontal StyledProps--main"
        >
          <div
            class="Layout--vertical Layout--layout-spacing-small StyledProps--main StyledProps--padding-right-large"
            style="width: 30%;"
          >
            <p
              class="StyledProps--font StyledProps--main"
            >
              common.city
            </p>
            <div
              class="TextInput--main"
            >
              <div
                class="bp3-input-group"
              >
                <input
                  class="bp3-input"
                  data-testid="city"
                  type="text"
                  value="dallas"
                />
              </div>
            </div>
          </div>
          <div
            class="Layout--horizontal StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-space-between StyledProps--flex-alignItems-baseline"
            style="width: 70%;"
          >
            <div
              class="Layout--vertical Layout--layout-spacing-small StyledProps--main"
            >
              <p
                class="StyledProps--font StyledProps--main"
              >
                common.state
              </p>
              <div
                class="bp3-popover-wrapper DropDown--main"
              >
                <div
                  class="bp3-popover-target"
                >
                  <div
                    class=""
                  >
                    <div
                      class="Layout--horizontal StyledProps--main DropDown--dropdownButton DropDown--withBorder DropDown--minWidth StyledProps--flex"
                      data-testid="state"
                      style="min-width: 130px;"
                    >
                      <div
                        class="Layout--horizontal StyledProps--main DropDown--labelWrapper StyledProps--flex"
                      >
                        <p
                          class="StyledProps--font Text--lineclamp StyledProps--main DropDown--label"
                          data-testid="dropdown-value"
                          style="--text-line-clamp: 1;"
                        >
                          Select
                        </p>
                      </div>
                      <div
                        class="Layout--horizontal StyledProps--main DropDown--btnWrapper StyledProps--flex"
                      >
                        <span
                          class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-grey400"
                          data-icon="main-chevron-down"
                        >
                          <svg
                            height="8"
                            viewBox="0 0 14 8"
                            width="8"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M13 0c-.3 0-.5.1-.7.3L7 5.6 1.7.3C1.5.1 1.3 0 1 0 .4 0 0 .4 0 1c0 .3.1.5.3.7l6 6c.2.2.4.3.7.3s.5-.1.7-.3l6-6c.2-.2.3-.4.3-.7 0-.6-.4-1-1-1z"
                              fill="#4d5969"
                            />
                          </svg>
                        </span>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="Layout--vertical Layout--layout-spacing-small StyledProps--main"
            >
              <p
                class="StyledProps--font StyledProps--main"
              >
                common.zipCode
              </p>
              <div
                class="TextInput--main"
              >
                <div
                  class="bp3-input-group"
                >
                  <input
                    class="bp3-input"
                    data-testid="zipCode"
                    type="text"
                    value="12345"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
